src/native/eventpipe/ep-session.c

```diff
@@ -523,6 +654,158 @@ ep_session_write_all_buffers_to_file (EventPipeSession *session, bool *events_wr
+		tracepoint = session_provider->default_tracepoint;
+	} else {
+		// We have a tracepoint for this event_id
+		tracepoint = dn_umap_it_value_t (found1, EventPipeTracepoint *);
+	}
+	if (tracepoint == NULL) {
+		// No tracepoint for this event_id and no default tracepoint, so we can't write the event.
+		return false;
+	}
+
+	if (tracepoint->enabled == 0) {
+		// No listeners
+		return false;
+	}
+
+	struct iovec io[9];
+
+	io[0].iov_base = &tracepoint->write_index;      // __u32 from event_reg
+	io[0].iov_len = sizeof(tracepoint->write_index);
+
+	uint8_t version = 0x01; // hardcoded for the first tracepoint format version
+	io[1].iov_base = &version;
+	io[1].iov_len = sizeof(version);
+
+	uint16_t truncated_event_id = event_id & 0xFFFF;
+	io[2].iov_base = &truncated_event_id;
+	io[2].iov_len = sizeof(truncated_event_id);
+
+	// The data transmitted in version 1 is
+	// extension - a NetTrace V6 LabelList
+	// payload - the EventPipe Event Payload
+	// meta - the EventPipe Event metadata
+
+	bool activity_id_is_empty = true;
+	if (activity_id != NULL) {
+		// If the activity_id is not empty, then we don't consider it empty.
+		for (int i = 0; i < EP_ACTIVITY_ID_SIZE; ++i) {
+			if (activity_id[i] != 0) {
+				activity_id_is_empty = false;
+				break;
+			}
+		}
+	}
+	bool related_activity_id_is_empty = true;
+	if (related_activity_id != NULL) {
+		// If the related_activity_id is not empty, then we don't consider it empty.
+		for (int i = 0; i < EP_ACTIVITY_ID_SIZE; ++i) {
+			if (related_activity_id[i] != 0) {
+				related_activity_id_is_empty = false;
+				break;
+			}
+		}
+	}
+	// extension generation helper
+	uint16_t extension_len = 0;
+	if (activity_id != NULL && !activity_id_is_empty)
+		extension_len += 1 + EP_ACTIVITY_ID_SIZE; // ActivityId kind + value
+	if (related_activity_id != NULL && !related_activity_id_is_empty)
+		extension_len += 1 + EP_ACTIVITY_ID_SIZE; // RelatedActivityId kind + value
+
+	uint8_t *extension = NULL;
+	if (extension_len > 0) {
+		extension = (uint8_t *)malloc(extension_len);
+		EP_ASSERT(extension != NULL);
+		uint16_t offset = 0;
+		if (activity_id != NULL && !activity_id_is_empty) {
+			// If there is a related_activity_id, use 0x81 (more follows), else 0x01 (no more follows)
+			extension[offset] = (related_activity_id != NULL && !related_activity_id_is_empty) ? 0x81 : 0x01;
+			memcpy(extension + offset + 1, activity_id, EP_ACTIVITY_ID_SIZE);
+			offset += 1 + EP_ACTIVITY_ID_SIZE;
+		}
+		if (related_activity_id != NULL && !related_activity_id_is_empty) {
+			// RelatedActivityId: 0x02 (no more follows)
+			extension[offset] = 0x02;
+			memcpy(extension + offset + 1, related_activity_id, EP_ACTIVITY_ID_SIZE);
+			offset += 1 + EP_ACTIVITY_ID_SIZE;
+		}
+	}
+
+	uint32_t payload_len = ep_event_payload_get_size (ep_event_payload);
+	if ((payload_len & 0xFFFF0000) != 0) {
+		// Payload is too large, we can't write it.
+		return false;
+	}
+	uint8_t *payload = (uint8_t *)malloc (payload_len);
```

Comment:
It doesn't need to be in this PR, but for performance we should avoid this allocate and copy. The logic inside ep_event_payload_copy_data is concatenating potentially many smaller buffers together into a temporary buffer. Rather than doing that we can just pass the pointers and sizes to writev() as yet more ranges for it to concatenate for us.

If you don't do it as part of this PR payload should be freed at the end of this method. Currently it is leaking.

Reply:
Thanks for the catch. After offline discussion, added a reasonably sized default static buffer to hold Event payloads that are stored in EventData. Refactored the event writing after review where the event payload is being written first, so in case an event payload has more EventData info than the static buffer, we dynamically allocate memory and then fill in the other known info. + freeing if we hit the dynamically allocated case